### PR TITLE
[jmx] add option to use cgroup memory limits for heap size

### PR DESCRIPTION
**Pull request by xvello**
Java 8u131 introduced an experimental `UseCGroupMemoryLimitForHeap` option, see these articles:

- https://blogs.oracle.com/java-platform-group/java-se-support-for-docker-cpu-and-memory-limits
- https://developers.redhat.com/blog/2017/04/04/openjdk-and-containers/

This PR introduces a `jmx_use_cgroup_memory_limit` boolean option (and its envvar binding), to enable this mechanism. This disables the default hardcoded `-Xmx200m -Xms50m` options we pass to jmxfetch.

In Agent5, [our default jmxfetch memory limits](https://github.com/DataDog/dd-agent/blob/dd512f5904c7c1ccd5d2267fa1d715e6999cc949/jmxfetch.py#L47-L48) were:
  - 512 MB if using autodiscovery
  - 200 MB if not

In Agent6 we went with 200 MB in both cases, which can be too conservative for heavy users.

This mechanism allows jmxfetch to scale down when the container memory limit is low (I tested it with 200 MB), and scale up when more memory is available:

Container Live view

### CODE REVIEW

**Review by CharlyF — DISMISSED** [review]

File: pkg/collector/corechecks/embed/jmx.go [hunk 1]
```diff
@@ -265,14 +272,24 @@ func (c *JMXCheck) start() error {
...
-		javaOptions += jvmDefaultMaxMemoryAllocation
-	}
-	// Specify the initial memory allocation pool for the JVM
-	if !strings.Contains(javaOptions, "Xms") && !strings.Contains(javaOptions, "XX:InitialHeapSize") {
-		javaOptions += jvmDefaultInitialMemoryAllocation
+	if config.Datadog.GetBool("jmx_use_cgroup_memory_limit") {
+		// This option is incompatible with the Xmx and Xms options, log a warning if there are found in the javaOptions
+		for _, option := range jvmCgroupMemoryIncompatOptions {
+			if strings.Contains(javaOptions, option) {
+				log.Warnf("Java option %q is incompatible with cgroup memory limit, please remove it", option)
```
Comment (truthbk) [inline pkg/collector/corechecks/embed/jmx.go]: Could we please explicitly state that the option it will be ignored.

**Review by truthbk — COMMENTED** [review]
This looks good to me 👍  

My only concern would be that we're applying the option blindly when set and I'm wondering if this could cause issues for users on certain JRE/JDKs that differ from the environments we have tested working. I know 99.9% of the times this is going to be used in an environment we control (ie. containers).... That said I can see customers maybe reusing configs or whatever, maybe mistakenly setting the option and it would probably be preferable if the process would still come up.

**Comment by xvello** [comment]
- older jre version fail to start jmxfetch, with an "unknown option -XX:+UseCGroupMemoryLimitForHeap" error, that we resurface in the agent.log and `agent status` output
- if a `Xmx`/`Xms` option is present, we currently still pass the CGroup option to java, which silently ignores it and runs as if the option was not set. Would you rather have us not send the option if we detect Xmx ?

File: pkg/collector/corechecks/embed/jmx.go (hunk 1 above)
Comment (xvello) [reply pkg/collector/corechecks/embed/jmx.go]: reworked the wording, and not passing the option, instead of passing it and having java silently ignore it

**Comment by xvello** [comment]
After testing, java 7 does not exit on unknown `XX` options, just silently ignores them:

```
$ docker run --rm -it -m 500000000 openjdk:8-slim java  -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap  -XX:+PrintFlagsFinal -version |grep MaxHeapSize
WARNING: Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.
    uintx MaxHeapSize                              := 132120576                           {product}
$ docker run --rm -it -m 500000000 openjdk:7-slim java  -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap  -XX:+PrintFlagsFinal -version |grep MaxHeapSize
WARNING: Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.
    uintx MaxHeapSize                              := 251658240       {product}
```

Java6 fails, but iirc we don't support it anymore, do we?
```
docker run --rm -it -m 500000000 openjdk:6-slim java  -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap  -XX:+PrintFlagsFinal -version
WARNING: Your kernel does not support swap limit capabilities or the cgroup is not mounted. Memory limited without swap.
Unrecognized VM option 'UseCGroupMemoryLimitForHeap'
Could not create the Java virtual machine.
```

**Review by CharlyF — APPROVED** [review]

**Review by olivielpeau — APPROVED** [review]
Documentation LGTM 👍

**Pull request merged**